### PR TITLE
Tiny tweaks and fixes for the PowerShell provisioner

### LINKED ISSUE #6188: filepath.[Dir|Base|ToSlash] behave differently depending on platform and path format
`filepath.Dir`, `filepath.Base`, and `filepath.ToSlash` behave differently depending on the platform they are used on and path formats fed to them e.g. forward or back slash separated paths. This is because they ultimately use `os.PathSeparator` to decide on how the provided path string should be split or manipulated. Clearly `os.PathSeparator` is different between platforms.

The basic problem can be demonstrated using [this script](https://gist.github.com/DanHam/449fc0bce0e5f33b6da3691baba11389)

On Windows this results in the following:
```
OS is windows
Path separator is: \

Forward path is C:/Windows/Temp/forward.txt
filepath.Dir(`C:/Windows/Temp/forward.txt`) gives: C:\Windows\Temp
filepath.Base(`C:/Windows/Temp/forward.txt`) gives: forward.txt

Forward path with spaces is C:/Windows/Temp/forward\ with\ space.txt
filepath.Dir(`C:/Windows/Temp/forward\ with\ space.txt`) gives: C:\Windows\Temp\forward\ with
filepath.Base(`C:/Windows/Temp/forward\ with\ space.txt`) gives:  space.txt

Backward path is C:\Windows\Temp\backward.txt
filepath.Dir(`C:\Windows\Temp\backward.txt`) gives: C:\Windows\Temp
filepath.Base(`C:\Windows\Temp\backward.txt`) gives: backward.txt

Converting the backward path with filepath.ToSlash(`C:\Windows\Temp\backward.txt`) gives: C:/Windows/Temp/backward.txt
For the converted backward path filepath.Dir(`C:/Windows/Temp/backward.txt`) gives C:\Windows\Temp
For the converted backward path filepath.Base(`C:/Windows/Temp/backward.txt`) gives backward.txt
```

On Linux/OS X this results in the following:

```
OS is linux
Path separator is: /

Forward path is C:/Windows/Temp/forward.txt
filepath.Dir(`C:/Windows/Temp/forward.txt`) gives: C:/Windows/Temp
filepath.Base(`C:/Windows/Temp/forward.txt`) gives: forward.txt

Forward path with spaces is C:/Windows/Temp/forward\ with\ space.txt
filepath.Dir(`C:/Windows/Temp/forward\ with\ space.txt`) gives: C:/Windows/Temp
filepath.Base(`C:/Windows/Temp/forward\ with\ space.txt`) gives: forward\ with\ space.txt

Backward path is C:\Windows\Temp\backward.txt
filepath.Dir(`C:\Windows\Temp\backward.txt`) gives: .
filepath.Base(`C:\Windows\Temp\backward.txt`) gives: C:\Windows\Temp\backward.txt

Converting the backward path with filepath.ToSlash(`C:\Windows\Temp\backward.txt`) gives: C:\Windows\Temp\backward.txt
For the converted backward path filepath.Dir(`C:\Windows\Temp\backward.txt`) gives .
For the converted backward path filepath.Base(`C:\Windows\Temp\backward.txt`) gives C:\Windows\Temp\backward.txt
```

### Path formatted with forward slashes:

* On Linux/OS X `filepath.Base` and `filepath.Dir` work as expected
* On Windows `filepath.Dir` converts the forward slashes to backward slashes. filepath.Base works as expected

### Path formatted with forward slashes including backslash escaped spaces

* On Linux/OS X `filepath.Base` and `filepath.Dir` work as expected
* On Windows since the path separator is `\`, `filepath.Dir` and `filepath.Base` split the path string on the last backslash which is not correct

### Path formatted with backward slashes

* On Linux/OS X since the path separator is `/`, `filepath.Base` returns the entire path rather than the filename. `filepath.Dir` returns a dot (`.`) as it cannot find any occurrences of the expected path separator in the string
* On Windows `filepath.Dir` and `filepath.Base` work as expected

### 'Converting' paths formatted with backward slashes to forward slashes

* On Linux/OS X since the path separator is `/`, `filepath.ToSlash` is a noop and the entire path is unchanged
* On Windows `filepath.ToSlash` works as expected. However, note that any `filepath.Dir` on the converted path will be affected by the issue noted above for paths formatted with forward slashes - namely that the forward slashes will be converted again to backward slashes. Again `filepath.Base` should work as expected for the converted path string.

## Example Issue

filepath.Dir, filepath.Base and filepath.ToSlash have been used throughout the code base. The use [here](https://github.com/hashicorp/packer/blob/master/communicator/ssh/communicator.go#L562-L571) in the SSH communicator was the root cause of one of the issues seen in #5963 as the PowerShell provisioner was using backslash formatted paths. On Linux/OS X this resulted in the upload path evaluating to `.` which ultimately caused the scp upload to fail. The issue was worked around for #5963 by configuring all of the paths in the PowerShell provisioner to use forward slashes. However, any user configuring a custom [remote_path](https://www.packer.io/docs/provisioners/powershell.html#remote_path) could reasonably expect to use a path formatted with back slashes. Doing so would unfortunately trigger the same issue.

It could be that this is the only example where the use of the filepath functions has caused an issue. However, executing a `grep -HRne filepath.[Dir|Base|ToSlash]` in the root of the Packer directory shows the extent to which these functions have been used; I wouldn't be surprised if there were a few other issues occurring because of this, especially where users can provide custom paths rather than using the internal defaults...

### PR DESCRIPTION
**Pull request by azr**
* make sure we're consistently documenting slashes
* fix a potential bug

fix #6188

### CODE REVIEW

File: provisioner/powershell/provisioner.go [hunk 1]
```diff
@@ -285,9 +286,9 @@ func (p *Provisioner) Provision(ctx context.Context, ui packersdk.Ui, comm packe
 		if err != nil {
 			return fmt.Errorf("Error stating powershell script: %s", err)
 		}
-		if strings.HasSuffix(p.config.RemotePath, `\`) {
+		if os.IsPathSeparator(p.config.RemotePath[len(p.config.RemotePath)-1]) {
 			// path is a directory
-			p.config.RemotePath += filepath.Base((fi).Name())
+			p.config.RemotePath += filepath.Base(fi.Name())
```
Comment (azr) [inline provisioner/powershell/provisioner.go:291]: `os.IsPathSeparator` on Windows will check for `\` and `/` ! `filepath.Base` internally uses `IsPathSeparator` to find the base path, so it should be fine.

**Review by nywilken — APPROVED** [review]
Overall this looks like a much needed change. There are two small updates needed for the documentation to wrap up the change. See comments inline. 

Approving in advance to unblock the merge.

File: website/content/docs/provisioners/powershell.mdx [hunk 2]
```diff
@@ -16,6 +16,9 @@ See the [section
 below](/docs/provisioners/powershell#combining-the-powershell-provisioner-with-the-ssh-communicator)
 for details.
 
+~> On Windows, if possible, try to always use a backslash `\` as a path
+separator, especially when dealing with non full paths. This is because when 
+
```
[the remark this answers is not in the document]
Comment (azr) [reply website/content/docs/provisioners/powershell.mdx:21]: Ha yes, good catch, this is a WIP commit I pushed before we switched over to the acc tests branch ! Fixing now.

[the remark this answers is not in the document]
Comment (azr) [reply website/content/docs/provisioners/powershell.mdx]: I made it clearer; and now wonder if we should display this in other places too. Ah, I will make a template out of this !

**Pull request merged**